controller_interface/include/semantic_components/semantic_component_interface.hpp

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2021, Stogl Robotics Consulting UG (haftungsbeschränkt)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef SEMANTIC_COMPONENTS__SEMANTIC_COMPONENT_INTERFACE_HPP_
+#define SEMANTIC_COMPONENTS__SEMANTIC_COMPONENT_INTERFACE_HPP_
+
+#include <string>
+#include <vector>
+
+#include "controller_interface/helpers.hpp"
+#include "hardware_interface/loaned_state_interface.hpp"
+
+namespace semantic_components
+{
+
+class SemanticComponentInterface
+{
+public:
+  explicit SemanticComponentInterface(const std::string & name, size_t size = 0)
+  {
+    name_ = name;
+    interface_names_.reserve(size);
+    state_interfaces_.reserve(size);
+  }
+
+  ~SemanticComponentInterface() = default;
+
+  /// Assign loaned state interfaces from the hardware.
+  /**
+   * Assign loaned state interfaces on the controller start.
+   *
+   * \param[in] state_interfaces vector of interfaces provided by the controller.
+   */
+  bool assign_loaned_state_interfaces(
+    std::vector<hardware_interface::LoanedStateInterface> & state_interfaces)
+  {
+    return controller_interface::get_ordered_interfaces(
+      state_interfaces, interface_names_, "", state_interfaces_);
+  }
+
+  /// Release loaned interfaces from the hardware.
+  void release_interfaces()
+  {
+    state_interfaces_.clear();
+  }
+
+  /// Definition of state interface names for the component.
+  /**
+   * The function should be used in "state_interface_configuration()" of a controller to provide
+   * standardized interface names semantic component.
+   *
+   * \default Default implementation defined state interfaces as "name/NR" where NR is number
+   * from 0 to size of values;
+   * \return list of strings with state interface names for the semantic component.
+   */
+  virtual std::vector<std::string> get_state_interface_names()
+  {
+    if (interface_names_.empty()) {
+      for (auto i = 0u; i < interface_names_.capacity(); ++i) {
+        interface_names_.emplace_back(name_ + "/" + std::to_string(i + 1));
+      }
+    }
+    return interface_names_;
+  }
+
+  /// Return all values.
+  /**
+   * \return vector with all stored values;
+   */
+  virtual std::vector<double> get_values() const
+  {
+    std::vector<double> values;
+    values.reserve(state_interfaces_.size());
+    for (size_t i = 0; i < state_interfaces_.size(); ++i) {
+      values.emplace_back(state_interfaces_[i].get().get_value());
+    }
+    return values;
+  }
```

Comment:
I'd add a `virtual void get_values(std::vector<double> & values) const` too. Otherwise there is no real-time safe way of accessing the data.

Reply:
@jordan-palacios , correct me if I am wrong,

You are asking something like,

```
virtual void get_values(std::vector<double> &values) const
{
    values.reserve(state_interfaces_.size());
    for (size_t i = 0; i < state_interfaces_.size(); ++i) {
        values.emplace_back(state_interfaces_[i].get().get_value());
    }
    return;
}
```
Is it my correct understanding? Pardon me, but can you please explain how this could be 
> real-time safe way of accessing the data.

Reply:
@bailaC: are you getting compilation issue here?

Reply:
I haven't tried it yet. Just wanted to make sure I understood the requirement here.

Reply:
@bailaC The current `get_values` creates a vector from scratch each time the function is called. When you reserve the space you will trigger a reallocation of the data. This most likely will end up in a system call (like `malloc`) that may reschedule the running thread. Anything that introduces unwanted latencies is considered not real-time safe.

By providing a `get_values` with a vector passed by reference the responsibility of maintaining the vector is left to the caller. 

Your sample code should work just fine!

Reply:
Thank you for the explanation. I have already done the changes and pushed it.